tools/slicec-cs/src/decoding.rs

```diff
@@ -397,6 +415,42 @@ fn decode_stream_parameter(type_ref: &TypeRef, namespace: &str, encoding: Encodi
     }
 }
 
+fn decode_result_field(type_ref: &TypeRef, namespace: &str, encoding: Encoding) -> CodeBlock {
+    if type_ref.is_optional {
+        let mut decode_func_body = decode_func_body(type_ref, namespace, encoding);
+
+        // TODO: it's lame to have to do this here. We should provide a better API.
+        if matches!(type_ref.concrete_type(), Types::Sequence(_) | Types::Dictionary(_)) {
+            write!(
+                decode_func_body,
+                " as {}",
+                type_ref.cs_type_string(namespace, TypeContext::Nested, false),
+            );
+        }
+
+        decode_func_body.indent();
+
+        // TODO: this won't result is a nice looking line when decode_func_body is multi-line
+        CodeBlock::from(format!(
+            "(ref SliceDecoder decoder) => decoder.DecodeBool() ? {decode_func_body} : null"
+        ))
+    } else {
+        let mut decode_func = decode_func(type_ref, namespace, encoding);
+
+        // Ditto
+        if matches!(type_ref.concrete_type(), Types::Sequence(_) | Types::Dictionary(_)) {
+            write!(
+                decode_func,
+                " as {}",
+                type_ref.cs_type_string(namespace, TypeContext::Nested, false),
+            );
+        }
+
+        decode_func.indent(); // TODO: this doesn't work
```

Comment:
What specifically doesn't look correct? We generally only call indent when something is about to be used as it could need to be indented multiple times depending on where it's being written.